test/libsolidity/NatspecJSONTest.cpp

```diff
@@ -0,0 +1,213 @@
+				"Malformed JSON in {} expectation for contract {}.\n"
+				"Note that JSON expectations must be pretty-printed to be split correctly. "
+				"The object is assumed to and at the first unindented closing brace.\n"
+				"{}",
+				toString(kind),
+				contractName,
+				rawJSON
+			)));
+
+		m_expectedNatspecJSON[string(contractName)][kind] = parsedJSON;
+	}
+}
+
+bool NatspecJSONTest::expectationsMatch()
+{
+	// NOTE: Comparing pretty printed Json::Values to avoid using its operator==, which fails to
+	// compare equal numbers as equal. For example, for 'version' field the value is sometimes int,
+	// sometimes uint and they compare as different even when both are 1.
+	return
+		SyntaxTest::expectationsMatch() &&
+		prettyPrinted(obtainedNatspec()) == prettyPrinted(m_expectedNatspecJSON);
+}
+
+void NatspecJSONTest::printExpectedResult(ostream& _stream, string const& _linePrefix, bool _formatted) const
+{
+	SyntaxTest::printExpectedResult(_stream, _linePrefix, _formatted);
+	if (!m_expectedNatspecJSON.empty())
+	{
+		_stream << _linePrefix << "----" << endl;
+		printIndented(_stream, formatNatspecExpectations(m_expectedNatspecJSON), _linePrefix);
+	}
+}
+
+void NatspecJSONTest::printObtainedResult(ostream& _stream, string const& _linePrefix, bool _formatted) const
+{
+	SyntaxTest::printObtainedResult(_stream, _linePrefix, _formatted);
+
+	NatspecMap natspecJSON = obtainedNatspec();
+	if (!natspecJSON.empty())
+	{
+		_stream << _linePrefix << "----" << endl;
+		// TODO: Diff both versions and highlight differences.
+		// We should have a helper for doing that in newly defined test cases without much effort.
+		printIndented(_stream, formatNatspecExpectations(natspecJSON), _linePrefix);
+	}
+}
+
+tuple<string_view, NatspecJSONKind> NatspecJSONTest::parseExpectationHeader(string_view _line)
+{
+	for (NatspecJSONKind kind: {NatspecJSONKind::Devdoc, NatspecJSONKind::Userdoc})
+	{
+		string kindSuffix = " " + toString(kind);
+		if (boost::algorithm::ends_with(_line, kindSuffix))
+			return {_line.substr(0, _line.size() - kindSuffix.size()), kind};
+	}
+
+	BOOST_THROW_EXCEPTION(runtime_error(
+		"Natspec kind (devdoc/userdoc) not present in the expectation: "s.append(_line)
+	));
+}
+
+string NatspecJSONTest::extractExpectationJSON(istream& _stream)
+{
+	string rawJSON;
+	string line;
+	while (getline(_stream, line))
+	{
+		string_view strippedLine = expectLinePrefix(line);
+		rawJSON += strippedLine;
+		rawJSON += "\n";
+
+		if (strippedLine.empty())
+			continue;
+
+		if (boost::algorithm::starts_with(strippedLine, "}"))
+			break;
+	}
+
+	return rawJSON;
+}
+
+string_view NatspecJSONTest::expectLinePrefix(string_view _line)
+{
+	size_t startPosition = 0;
+	if (!boost::algorithm::starts_with(_line, "//"))
+		BOOST_THROW_EXCEPTION(runtime_error(
+			"Expectation line is not a comment: "s.append(_line)
+		));
+
+	startPosition += 2;
+	if (startPosition < _line.size() && _line[startPosition] == ' ')
```

Comment:
Should this behave as a `trim` function, or do we always expect a maximum of one whitespace after `//`?

Reply:
No, I intentionally only strip `// `. If there's more indentation there, it's a part of the value. In many cases it won't matter, but in some it might. I'm considering moving this helper to some more general location later, so better not to assume that. Actually, even with JSON it might matter if the parser allows multi-line strings (which I think jsoncpp does). And if you want to print the extracted value, it may look weird without whitespace.